packages/fork-choice/src/forkChoice/forkChoice.ts

```diff
@@ -921,11 +904,32 @@ export class ForkChoice implements IForkChoice {
   }
 }
 
-function assertValidTerminalPowBlock(config: IChainConfig, block: PowBlockHex, parent: PowBlockHex): void {
-  const isTotalDifficultyReached = block.totalDifficulty >= config.TERMINAL_TOTAL_DIFFICULTY;
-  const isParentTotalDifficultyValid = parent.totalDifficulty < config.TERMINAL_TOTAL_DIFFICULTY;
+function assertValidTerminalPowBlock(
+  config: IChainConfig,
+  block: merge.BeaconBlock,
+  preCachedData?: OnBlockPrecachedData
+): void {
+  const {powBlock, powBlockParent} = preCachedData || {};
+  if (!powBlock) throw Error("onBlock preCachedData must include powBlock");
+  if (!powBlockParent) throw Error("onBlock preCachedData must include powBlock");
+
+  if (!ssz.Root.equals(config.TERMINAL_BLOCK_HASH, ZERO_HASH)) {
+    if (computeEpochAtSlot(block.slot) < config.TERMINAL_BLOCK_HASH_ACTIVATION_EPOCH)
+      throw Error(`Terminal block activation epoch ${config.TERMINAL_BLOCK_HASH_ACTIVATION_EPOCH} not reached`);
+
+    // powBock.blockhash is hex, so we just pick the corresponding root
+    if (!ssz.Root.equals(block.body.executionPayload.parentHash, config.TERMINAL_BLOCK_HASH))
+      throw new Error(
+        `Invalid terminal block hash, expected: ${toHexString(config.TERMINAL_BLOCK_HASH)}, actual: ${toHexString(
+          block.body.executionPayload.parentHash
+        )}`
+      );
+  }
```

Comment:
This code still have the same issue:
```ts
if (!ssz.Root.equals(config.TERMINAL_BLOCK_HASH, ZERO_HASH)) {
  // block hash checks
} else {
  // total difficulty checks
}
```

Reply:
umm... there is a total difficulty test on L929:934, irrespective if `TERMINAL_BLOCK_HASH` is non zero or not.

Reply:
Please review the spec code, after the condition `if TERMINAL_BLOCK_HASH != Hash32():` the block ends with a `return` so the total difficulty checks are **NOT** run if `TERMINAL_BLOCK_HASH` is defined

https://github.com/ethereum/consensus-specs/blob/dev/specs/merge/fork-choice.md#validate_merge_block

Reply:
yea totally missed return, updating commit